### ISSUE TITLE
Modelica-Buildings Library - model fails due to the array dimension is assigned by using the parameter start attribute

### ISSUE THREAD

**Issue by AndreaBartolini** [OpenModelica/OpenModelica#12305]
### Description
The following models in the Buildings Library:
```
Buildings.Templates.Plants.HeatPumps.Components.Validation.HeatPumpGroupAirToWater
Buildings.Templates.Plants.HeatPumps.Components.Validation.PumpsPrimaryDedicated
Buildings.Templates.Plants.HeatPumps.Components.Validation.ValvesIsolation
```
fail to simulate in OpenModelica by giving the following error:
`Error: Could not evaluate structural parameter (or constant): ctlPlaAwNrv.nAirHan which gives dimensions of array: busAirHan. Array dimensions must be known at compile time.`

https://libraries.openmodelica.org/branches/master/Buildings_latest/files/Buildings_latest_Buildings.Templates.Plants.HeatPumps.Components.Validation.HeatPumpGroupAirToWater.err

This because in that models the value of the `nAirHan` parameter is not assigned by using modifiers so it should be determined by using its start attribute:
```modelica
  parameter Integer nAirHan(
    final min=if typ <> Buildings.Templates.Plants.HeatPumps.Types.Controller.OpenLoop
      and nEquZon == 0 then 1 else 0,
    start=0)
```

Apparently OpenModelica does not use the `start` attribute of the parameter to evaluate this value in order to determine the value of the `busAirHan` array.

A simple MWE to reproduce this issue is given in #12304 .

### Version and OS
https://libraries.openmodelica.org/branches/master/Buildings_latest/Buildings_latest.html

**Comment by AndreaBartolini** [comment]
keep in in the loop:
@deepak19015 @casella

**Comment by casella** [comment]
I would suggest to open a ticket for the Buildings library and ask them to explicitly add a modifier on the example model. Start attributes are only meant for toy models and will produce a warning.

**Comment by AndreaBartolini** [comment]
As far as I sow this modus-operandi is wide used in the Buildings Library, maybe this should be discussed with them at more general level.

**Comment by AndreaBartolini** [comment]
BTW, the problem is solved for the scalar case but not for the arrays... see https://github.com/OpenModelica/OpenModelica/issues/12304

**Comment by casella** [comment]
@perost, #12304 is now fixed, but the models

[Buildings.Templates.Plants.HeatPumps.Components.Validation.HeatPumpGroupAirToWater](https://libraries.openmodelica.org/branches/master/Buildings_latest/files/Buildings_latest_Buildings.Templates.Plants.HeatPumps.Components.Validation.HeatPumpGroupAirToWater)
[Buildings.Templates.ZoneEquipment.Validation.VAVBoxCoolingOnly](
https://libraries.openmodelica.org/branches/master/Buildings_latest/files/Buildings_latest_Buildings.Templates.ZoneEquipment.Validation.VAVBoxCoolingOnly.err)
[Buildings.Templates.ZoneEquipment.Validation.VAVBoxReheat](https://libraries.openmodelica.org/branches/master/Buildings_latest/files/Buildings_latest_Buildings.Templates.ZoneEquipment.Validation.VAVBoxReheat.err
)
[Buildings.Templates.Plants.HeatPumps.Components.Validation.ValvesIsolation](Buildings.Templates.Plants.HeatPumps.Components.Validation.ValvesIsolation)
[Buildings.Templates.Plants.HeatPumps.Validation.AirToWater.err](https://libraries.openmodelica.org/branches/master/Buildings_latest/files/Buildings_latest_Buildings.Templates.Plants.HeatPumps.Validation.AirToWater.err)

still fail during flattening. Not sure if this is related, can you please check? Thanks!

**Comment by casella** [comment]
@perost if you can have a quick look at this, that would be great. Thanks!

**Comment by casella** [comment]
@perost ping 😃

**Comment by perost** [comment]
> @perost, #12304 is now fixed, but the models
> 
> [Buildings.Templates.Plants.HeatPumps.Components.Validation.HeatPumpGroupAirToWater](https://libraries.openmodelica.org/branches/master/Buildings_latest/files/Buildings_latest_Buildings.Templates.Plants.HeatPumps.Components.Validation.HeatPumpGroupAirToWater) [Buildings.Templates.ZoneEquipment.Validation.VAVBoxCoolingOnly](https://libraries.openmodelica.org/branches/master/Buildings_latest/files/Buildings_latest_Buildings.Templates.ZoneEquipment.Validation.VAVBoxCoolingOnly.err) [Buildings.Templates.ZoneEquipment.Validation.VAVBoxReheat](https://libraries.openmodelica.org/branches/master/Buildings_latest/files/Buildings_latest_Buildings.Templates.ZoneEquipment.Validation.VAVBoxReheat.err) [Buildings.Templates.Plants.HeatPumps.Components.Validation.ValvesIsolation](Buildings.Templates.Plants.HeatPumps.Components.Validation.ValvesIsolation) [Buildings.Templates.Plants.HeatPumps.Validation.AirToWater.err](https://libraries.openmodelica.org/branches/master/Buildings_latest/files/Buildings_latest_Buildings.Templates.Plants.HeatPumps.Validation.AirToWater.err)
> 
> still fail during flattening. Not sure if this is related, can you please check? Thanks!

I had a look at the `VAVBoxReheat` model, and got the following MWE:
```Modelica
record DXCoil
  parameter Integer nSta = 1;
  parameter Real sta[nSta];
end DXCoil;

record CoilData
  parameter Boolean have_sou;
  parameter DXCoil datCoi;
end CoilData;

record VAVBoxData
  parameter CoilData coiHea;
end VAVBoxData;

model M
  parameter CoilData coilData;
  VAVBoxData data(coiHea = coilData);
end M;
```
In this case `nSta` obviously has a value, but the `have_sou` parameter doesn't and somehow causes the evaluation of `nSta` to fail. I don't know if this is also the case with the full model, it might be that it just happens to give the same error. But I'll need to fix this MWE so that it at least gives the correct error before I can debug the full model further if needed.

**Comment by perost** [comment]
#13024 fixes the "Could not evaluate structural parameter " issue. With this fix the model instead fails with no error message, which turned out to be due to issues with expandable connectors.

I'll be working on expandable connectors in the coming weeks, because the current implementation has a lot of issues. But I made a quick fix in #13025 that at least makes it print some error messages that indicate what the issue is rather than just failing silently.

**Comment by casella** [comment]
Original issues fixed.
